x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/mr/ItemSetMapReduceValueSource.java

```diff
@@ -150,51 +161,211 @@ Field getField() {
 
-                    if (valuesCount == 0) {
-                        return empty;
+            GlobalOrdinals(
+                Field field,
+                Bytes.WithOrdinals source,
+                IncludeExclude.OrdinalsFilter globalOrdinalsFilter,
+                LeafReaderContext ctx
+            ) throws IOException {
+                this.field = field;
+                this.source = source;
+
+                bitSetFilter = globalOrdinalsFilter != null
+                    ? globalOrdinalsFilter.acceptedGlobalOrdinals(source.globalOrdinalsValues(ctx))
+                    : null;
+
+                this.docValues = source.globalOrdinalsValues(ctx);
+            }
+
+            @Override
+            public ValueCollector getValueCollector(LeafReaderContext ctx) throws IOException {
+                final SortedSetDocValues values = source.globalOrdinalsValues(ctx);
+                final Tuple<Field, List<Object>> empty = new Tuple<>(field, Collections.emptyList());
+
+                return doc -> {
+                    if (values.advanceExact(doc)) {
+                        int valuesCount = values.docValueCount();
+
+                        if (valuesCount == 1) {
+                            long v = values.nextOrd();
+                            assert v >= 0;
+                            if (bitSetFilter == null || bitSetFilter.get(v)) {
+                                return new Tuple<>(field, Collections.singletonList(v));
+                            }
+                            return empty;
+                        }
+
+                        if (valuesCount == 0) {
+                            return empty;
+                        }
+
+                        List<Object> objects = new ArrayList<>(valuesCount);
+
+                        for (int i = 0; i < valuesCount; ++i) {
+                            long v = values.nextOrd();
+                            assert v >= 0;
+                            if (bitSetFilter == null || bitSetFilter.get(v)) {
+                                objects.add(v);
+                            }
+                        }
+                        return new Tuple<>(field, objects);
                     }
+                    return empty;
+                };
+            }
 
-                    List<Object> objects = new ArrayList<>(valuesCount);
+            @Override
+            public boolean usesOrdinals() {
+                return true;
+            }
 
-                    for (int i = 0; i < valuesCount; ++i) {
-                        BytesRef v = values.nextValue();
-                        if (stringFilter == null || stringFilter.accept(v)) {
-                            objects.add(BytesRef.deepCopyOf(v));
+            @Override
+            public Object mapOrdinal(Object ord) throws IOException {
+                return docValues.lookupOrd((Long) ord);
+            }
+        }
+
+        static class Map implements ExecutionStrategy {
```

Comment:
Calling this class `Map` might cause people to suffer editor autocomplete or navigation issues. Maybe `MapStrategy`? And call the other class `GlobalOrdinalsStrategy`?